## Day 11 Assignment : 
We are to make Partial Correlations between ***Attrition*** and 10 other numerical features in the dataset and frame H0 and H1 hypothesis statements for approving or disapporving either one. Also we are going to make scatterplots to indicate a -ve, 0 or +ve correlations. Given Significance Level = 0.05

### Step 1 : 
Importing the libraries and the dataset

In [4]:
# Importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [5]:
# Importing the dataset
dt = pd.read_csv("general_data.csv")
dt.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

### Step 2 : 
Data Preprocessing Phase 

In [6]:
# We check for null values in our data
dt.isnull()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  \
0     False      False           False       False             False   
1     False      False           False       False             False   
2     False      False           False       False             False   
3     False      False           False       False             False   
4     False      False           False       False             False   
...     ...        ...             ...         ...               ...   
4405  False      False           False       False             False   
4406  False      False           False       False             False   
4407  False      False           False       False             False   
4408  False      False           False       False             False   
4409  False      False           False       False             False   

      Education  EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0         False           False          False       False   False  ...   
1         False           False          False       False   False  ...   
2         False           False          False       False   False  ...   
3         False           False          False       False   False  ...   
4         False           False          False       False   False  ...   
...         ...             ...            ...         ...     ...  ...   
4405      False           False          False       False   False  ...   
4406      False           False          False       False   False  ...   
4407      False           False          False       False   False  ...   
4408      False           False          False       False   False  ...   
4409      False           False          False       False   False  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
0                  False   False              False          False   
1                  False   False              False          False   
2                  False   False              False          False   
3                  False   False              False          False   
4                  False   False              False          False   
...                  ...     ...                ...            ...   
4405               False   False              False          False   
4406               False   False              False          False   
4407               False   False              False          False   
4408               False   False              False          False   
4409               False   False              False          False   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                False              False                  False   
1                False              False                  False   
2                False              False                  False   
3                False              False                  False   
4                False              False                  False   
...                ...                ...                    ...   
4405             False              False                  False   
4406             False              False                  False   
4407             False              False                  False   
4408             False              False                  False   
4409             False               True                  False   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0              False                    False                 False  
1              False                    False                 False  
2              False                    False                 False  
3              False                    False                 False  
4              False                    False                 False  
...              ...                      ...                   ...  
4405           False                    False                 False  
4406           False    

In [7]:
# We check for duplicates
dt.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 4410, dtype: bool

In [8]:
# Now we remove the null and the duplicated values
dt1 = dt.dropna()

In [9]:
dt2 = dt1.drop_duplicates()
dt2

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4404   29        No      Travel_Rarely                   Sales   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4404                 4          3          Other              1        4405   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4404  Female  ...                 1.0      Y                18              8   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4404                 0               6.0                      2   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

### Step 3 : 
First we convert ***Attrition*** which is a categorical variable into numerical i.e, 1s and 0s

In [10]:
attri = dt2.Attrition
attri

0        No
1       Yes
2        No
3        No
4        No
       ... 
4404     No
4405     No
4406     No
4407     No
4408     No
Name: Attrition, Length: 4382, dtype: object

As we can see the ***Attrition*** values are in **YES** and  **NO** . So we use the `cat.codes` method to convert them into **1**s and **0**s

In [11]:
dt2["Attrition"] = dt2["Attrition"].astype('category')
dt2["Attrition"] = dt2["Attrition"].cat.codes

Now we check to see the values of ***Attrition***

In [13]:
attri = dt2.Attrition
attri

0       0
1       1
2       0
3       0
4       0
       ..
4404    0
4405    0
4406    0
4407    0
4408    0
Name: Attrition, Length: 4382, dtype: int8

As we can see the **YES** and **NO** values have changed into **0**s and **1**s

### Step 4 :
Now we proceed with the **Correlation** :

First we import the library `pearsonr` from `scipy.stats` that is to be used for correlation.

In [15]:
from scipy.stats import pearsonr

1. Checking correlation between  ***Attrition*** and ***Age***

In [28]:
stats,p = pearsonr(dt2.Attrition, dt2.Age)
stats, p

(-0.15839867954096146, 5.126598219398078e-26)

As we can see the p-value `5.126598219398078e-26<<<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***Age*** 

2. Checking correlation between  ***Attrition*** and ***DistanceFromHome***

In [27]:
stats,p = pearsonr(dt2.Attrition, dt2.DistanceFromHome)
stats, p

(-0.009448638515156003, 0.5317715668019634)

As we can see the p-value `0.5317715668019634>0.05` so we accept the null hypothesis.
Therefore there is no significant correaltion between ***Attrition*** and ***DistanceFromHome*** 

3. Checking correlation between  ***Attrition*** and ***Education***

In [29]:
stats,p = pearsonr(dt2.Attrition, dt2.Education)
stats, p

(-0.017106307050278116, 0.2575753930816995)

As we can see the p-value `0.2575753930816995>0.05` so we accept the null hypothesis.
Therefore there is no significant correaltion between ***Attrition*** and ***Education*** 

4. Checking correlation between  ***Attrition*** and  ***PercentSalaryHike***

In [30]:
stats,p = pearsonr(dt2.Attrition, dt2.PercentSalaryHike)
stats, p

(0.03315303713546523, 0.028192446935107012)

As we can see the p-value `0.028192446935107012<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***PercentSalaryHike*** 

5. Checking correlation between  ***Attrition*** and ***TotalWorkingYears***

In [21]:
stats,p = pearsonr(dt2.Attrition, dt2.TotalWorkingYears)
stats, p

(-0.1696699168472327, 1.1645434967153252e-29)

As we can see the p-value `1.1645434967153252e-29<<<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***TotalWorkingYears*** 

 6. Checking correlation between  ***Attrition*** and ***NumCompaniesWorked***

In [31]:
stats,p = pearsonr(dt2.Attrition, dt2.NumCompaniesWorked)
stats, p

(0.04283056724471901, 0.004572057121625154)

As we can see the p-value `0.004572057121625154<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***NumCompaniesWorked*** 

 7. Checking correlation between  ***Attrition*** and ***YearsWithCurrManager***

In [23]:
stats,p = pearsonr(dt2.Attrition, dt2.YearsWithCurrManager)
stats, p

(-0.154691536902868, 7.1053696467956645e-25)

As we can see the p-value `7.1053696467956645e-25<<<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***YearsWithCurrManager*** 

 8. Checking correlation between  ***Attrition*** and ***YearSinceLastPromotion***

In [24]:
stats,p = pearsonr(dt2.Attrition, dt2.YearsSinceLastPromotion)
stats, p

(-0.03142315056330794, 0.03752293607393637)

As we can see the p-value ` 0.03752293607393637<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***YearsSinceLastPromotion*** 

 9. Checking correlation between  ***Attrition*** and ***YearsAtCompany***

In [25]:
stats,p = pearsonr(dt2.Attrition, dt2.YearsAtCompany)
stats, p

(-0.13300261842521083, 9.476118084864852e-19)

As we can see the p-value `9.476118084864852e-19<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***YearsSinceLastPromotion*** 

10. Checking correlation between  ***Attrition*** and ***TrainingTimeLastYear***

In [26]:
stats,p = pearsonr(dt2.Attrition, dt2.TrainingTimesLastYear)
stats, p

(-0.047585736930815525, 0.0016276603635481809)

As we can see the p-value `0.0016276603635481809<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition*** and ***TrainingTimeLastYear*** 